Heliwizard & Lua Swash access

jrwieland opened the pull request:
Heli wizard (for horus "color" based radio's) 
works for both FBL and flybar helicopters.

Added lua script access swash settings in to Model Edit for the fbl settings to be applied to the heli menu from the script

jrwieland committed:
Swash settings added for heli wizard

added LUA access to swash settings to allow the heli wizard(s) to set up required sources and values for the flybar equipped helicopters.

jrwieland committed:
heliwizard

The long awaited heli wizard for color radios 
taranis x7 & x9 heli wizards coming in the near future

jrwieland committed:
Delete wizard.luac

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

3djc commented:
Very interesting ! Has there been community test and discussion about this ? If yes, a link would be appreciated

jrwieland commented:
> 
> 
> Very interesting ! Has there been community test and discussion about this ? If yes, a link would be appreciated

@3djc  Here is a link for the files, there is a compiled companion (windows) as well as the C++ modeledit file and the lua script.
If you have another OS feel free to add the C++ file to it and compile a copy for yourself.
Poke around and see what needs changed... having more eyes on it than mine is always a plus !!

[Heli Test Files](https://drive.google.com/file/d/1Y0NOxN2OI29ajCTlWgCys6wGeZq7SU90/view?usp=sharing)

3djc commented on radio/src/lua/api_model.cpp:
```diff
@@ -1586,11 +1586,69 @@ static int luaModelResetSensor(lua_State *L)
+
+Set heli swash parameters
+
+index = 0
```
Not sure what this means ? What does index param correspond to  ?

jrwieland replied:
Parameter renamed and explained.

Since this is basically an internal mix it follows the same format as a mix
index refereed to "which" mix (swash) and Line refereed to what line.

Since this is the only mix line for this item I simply described it so the user no option other than to use "0" for these 2 items.

I have tried changing the values to 1 & 2, from 0 they have no effect, however to eliminate any possible conflicts I suggest they be set to "0" in the descripton.

![doc changes](https://user-images.githubusercontent.com/20526716/125211148-6430c900-e272-11eb-95eb-34e8e7d09b2f.PNG)

3djc commented on radio/sdcard/horus/SCRIPTS/WIZARD/heli/wizard.lua:
```diff
@@ -0,0 +1,559 @@
+    elseif TypeFields[2][5]==1 then
+      lcd.drawText(290,10,"Swash 120X")
+    elseif TypeFields[2][5]==2 then
+      lcd.drawText(290,10,"Swash 140") else
```
your "else" position and indentation needs to be changed (many instances of that)

jrwieland replied:
else position and indentation changed

3djc commented on radio/src/lua/api_model.cpp:
```diff
@@ -1586,11 +1586,69 @@ static int luaModelResetSensor(lua_State *L)
+unsigned int idx = luaL_checkunsigned(L, 1);
+if (idx < MIXSRC_CYC3){
+  luaL_checktype(L, -1, LUA_TTABLE);
+for (lua_pushnil(L); lua_next(L, -2); lua_pop(L, 1)) {
```
wrong indent

jrwieland replied:
indentation repaired

3djc commented on radio/src/lua/api_model.cpp:
```diff
@@ -1586,11 +1586,69 @@ static int luaModelResetSensor(lua_State *L)
+static int luaModelSwashRingData(lua_State *L)
+{
+unsigned int idx = luaL_checkunsigned(L, 1);
+if (idx < MIXSRC_CYC3){
```
You seemt to be checking idx and not using it ?

jrwieland replied:
unused idx removed,
it was added to reflect the other lua set commands format.
I didn't think this would work without an initial "if" statement ... I stand corrected. 

That's what I get for not coding C in over 35 years... second guessing myself and adding unnecessary items

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Swash indent cleanup

Unused idx removed (api-model)
Lua Document use explanation repaired
Indentation repaired (api_mode & wizard)
else command location changed (wizard)
checked wizard file (white space added for readability)

3djc commented on radio/sdcard/horus/SCRIPTS/WIZARD/heli/wizard.lua:
```diff
@@ -0,0 +1,592 @@
+    model.swashRingData(0,0,{type="3",collectiveSource=89,aileronSource=90,elevatorSource=88,collectiveWeight=40,aileronWeight=40,elevatorWeight=60})
+  elseif TypeFields[2][5]==2 then
+    model.swashRingData(0,0,{type="4",collectiveSource=89,aileronSource=90,elevatorSource=88,collectiveWeight=35,aileronWeight=35,elevatorWeight=60})
+  else
```
Something is missing or too much

jrwieland replied:
Without the else end (do nothing) , it always adds info to the heli screen.  This information is not needed for FBL helicopters, this screen pertains to flybar helicopters.

instead of a "do nothing" at the end I can move it to the beginning

if TypeFields[1][5]==0 then 
end
else
if TypeFields[2][5]==0 then 
...
else
model.swashRingData(0,0,{type="4" ....
end

It changes the location of the "do nothing" so that the command ends with a flybar choice.

Which method is preferred?

jrwieland replied:
Changed the location of the do nothing  to the front of the command - unwanted results.
With an "else statement" at the end, the swash parameters are always filled in regardless of helicopter type - flybarless or flybar.  (Since it didn't meet any of the other requirements it chose the last one).

Item reverted back to original setting with do nothing at the end.

3djc commented on radio/src/lua/api_model.cpp:
```diff
@@ -1586,11 +1586,68 @@ static int luaModelResetSensor(lua_State *L)
+
+Set heli swash parameters
+
+@param swash (unsigned number) number (use 0)
```
If it needs to be 0, does it make sense to ask inputing it ?

jrwieland replied:
No, but it has to be in the command otherwise it will throw an error when executing it, a similar situation with the line number, so there has to be an explanation.

Would something like
@param swash  must be zero "0"

be acceptable?

jrwieland replied:
language changed to required
![Capture](https://user-images.githubusercontent.com/20526716/125503092-cf11dae4-e3a8-400d-8958-07b7530eda7e.PNG)

jrwieland committed:
Swash Document Updated

unassigned parameter number for swash and line changed to required

3djc commented:
@jrwieland there was a number of issues, making it difficult to comment line by line. I have created a branch here : https://github.com/opentx/opentx/tree/3djc/lua-swash

Could you try it and let me know what you think ?

jrwieland committed:
Horus Heli Wizard

jrwieland commented:
files updated

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Heli wizards added

Added x9 Heli wizard.

Added x7 Heli wizard

x9 Clutter cleanup

Changed the directory structure for the x9  wizard (added "img" folder) .

Moved all the images to the folder.

Relocated all individual x9 wizards to the "img" folder .

Redirected the wizard to look in the img folder for corresponding wizards.

Added img/ to each drawPixmap command in the existing x9 wizards so their function/structure was preserved.

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

jrwieland commented:
@3djc Just wondering if you had a chance to recheck the heli Lua so it can make the 2.3.15 update when it comes out. 
I see that the edgeTX team  is looking to incorporate the code and tweak the wizard to work with the changes to their code

jrwieland committed:
Merge branch 'opentx:2.3' into 2.3

bsongis commented:
I merge it, but I propose those changes:
- rename the new function to `setSwashRing` 
- add a function `getSwashRing`

jrwieland commented:
Thank you for merging it.  Any changes you believe would be beneficial are welcomed

bsongis commented:
`swashRingData` renamed to `setSwashRing` in origin/2.3

3djc commented:
Could you do a new PR to add a getSwashRing function (so we have the get/set pair). The other piece has been done already

jrwieland commented:
Can do.. it'll be a little later today..that work thing is interfering with my fun time

3djc commented:
No rush ! Thanks for your help actually !
